File: src/main/GO/leetcode/lists/offer/Offer13.go
package offer

import "container/list"

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/4 16:06
 */

func movingCount(m int, n int, k int) int {
	queue := list.List{}
	visited := make([][]bool, m)
	for i := 0; i < len(visited); i++ {
		visited[i] = make([]bool, n)
	}
	queue.PushBack([]int{0, 0, 0, 0})
	res := 0
	for queue.Len() > 0 {
		back := queue.Back()
		queue.Remove(back)
		bfs := back.Value.([]int)
		i := bfs[0]
		j := bfs[1]
		si := bfs[2]
		sj := bfs[3]
		if i >= m || j >= n || si+sj > k || visited[i][j] {
			continue
		}
		res++
		visited[i][j] = true

		sj1 := sj + 1
		si1 := si + 1
		if (j+1)%10 == 0 {
			sj1 = sj - 8
		}
		if (i+1)%10 == 0 {
			si1 = si - 8
		}
		queue.PushBack([]int{i + 1, j, si1, sj})
		queue.PushBack([]int{i, j + 1, si, sj1})
	}
	return res
}

File: java/src/leetcode/lists/offer/Offer13.java
package leetcode.lists.offer;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/9/12 15:11
 */
public class Offer13 {

    public static void main(String[] args) {
        System.out.println(new Offer13().cal(1, 2));
    }

    public int movingCount(int m, int n, int k) {
        int[][] dic = new int[][]{{-1, 0}, {1, 0}, {0, 1}, {0, -1}};
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{0, 0});
        HashSet<String> set = new HashSet<>();
        set.add(0 + "," + 0);
        while (!queue.isEmpty()) {
            int[] node = queue.poll();
            for (int[] ints : dic) {
                int nr = node[0] + ints[0];
                int nc = node[1] + ints[1];
                if (nr < 0 || nc < 0 || nr >= m || nc >= n || cal(nr, nc) > k || set.contains(nr + "," + nc)) {
                    continue;
                }
                set.add(nr + "," + nc);
                queue.add(new int[]{nr, nc});
            }
        }
        return set.size();
    }

    public int cal(int i, int j) {
        return i % 10 + i / 10 + j % 10 + j / 10;
    }
}
